feat: memoize build function

eduardoboucas opened the pull request:
**- Summary**

Use memoization instead of debouncing the build function.

**- Test plan**

Reinstates the TypeScript tests.

**- A picture of a cute animal (not mandatory but encouraged)**

![cat-sample_1313](https://user-images.githubusercontent.com/4162329/116383540-d74ba700-a80e-11eb-87d1-9aacc0f0b147.jpg)

erezrokah commented on src/utils/serve-functions.js:
```diff
@@ -515,21 +515,35 @@ const setupFunctionsBuilder = async ({ config, errorExit, functionsDirectory, lo
+      )
+    }
+
+    return cache.get(key)
```
I couldn't find where the return value is used

eduardoboucas replied:
It's not used currently, but `debouncedBuild` preserved the return value of the build function and I thought `memoizedBuild` should respect that.

erezrokah commented on src/utils/serve-functions.js:
```diff
@@ -515,21 +515,35 @@ const setupFunctionsBuilder = async ({ config, errorExit, functionsDirectory, lo
+
+  // `memoizedBuild` will avoid consecutive calls to the build function with
+  // the same arguments until the previous invokation has finished running.
+  const memoizedBuild = (path, eventType) => {
```
If the user edits the same file multiple times only the first triggers a build?
Do we want to run the last one too?

erezrokah reviewed (changes requested):
Thanks @eduardoboucas, I think we might need a combination of debounce and memoize

ehmicky commented on src/utils/serve-functions.js:
```diff
@@ -515,21 +515,35 @@ const setupFunctionsBuilder = async ({ config, errorExit, functionsDirectory, lo
+    if (!cache.has(key)) {
+      cache.set(
+        key,
+        buildFunction(path, eventType).finally(() => cache.delete(key)),
```
Would using an async-aware memoization library like [`moize`](https://github.com/planttheidea/moize) work here?
https://github.com/planttheidea/moize#ispromise

eduardoboucas commented:
In https://github.com/netlify/cli/pull/2258/commits/b9f99f49130755c5d431bda31a3a00a16b10f1de I've updated the `memoizeBuild` to ensure any updates performed while the function builder is running are executed afterwards, whilst at the same time being able to discard any duplicate filesystem events in the same way that `debounce` used to do.

The implementation is non-trivial, but hopefully the comments I've added help clarify the behaviour.

> If the user edits the same file multiple times only the first triggers a build?
> Do we want to run the last one too?

@erezrokah This should now be addressed.

> Would using an async-aware memoization library like moize work here?

I think this is a special case where we want a time-based memoization with a debounce interval. I'm not aware of any library that does exactly what we need, but I'm happy to explore alternatives.
